File: shell.d/platform-freebsd/network-interfaces.rb
#!/usr/bin/env ruby

def cidr_to_mask(cidr)
  bits = (("1"*cidr.to_i) + ("0"*(32-cidr.to_i)))

  return [
    bits[0..7].to_i(2),
    bits[8..15].to_i(2),
    bits[16..23].to_i(2),
    bits[24..31].to_i(2),
  ].join('.')
end

interface_names = []

# get all interface names
interface_names = %x{ ifconfig -l }.strip.chomp.split(/\s+/).uniq

global_addresses = []

# for each interface...
interface_names.each_index do |i|
  interface = interface_names[i]
  addresses = []
  info = {}

# get interface details
  IO.popen("ifconfig '#{ interface }'").read.split("\n").each do |line|
    line = line.strip.chomp

    case line
    when /^#{ interface }:\s+flags=\d+<([^>]+)> metric (\d+) mtu (\d+)$/
      flags = $~[1].to_s.split(',').collect{|i| i.downcase }
      metric = $~[2].to_i
      mtu = $~[3].to_i

      puts "network.interfaces.#{ i }.name:str:#{ interface }"
      puts "network.interfaces.#{ i }.metric:int:#{ metric }"
      puts "network.interfaces.#{ i }.mtu:int:#{ mtu }"

      flags.each_index do |fi|
        puts "network.interfaces.#{ i }.flags.#{ fi }:str:#{ flags[fi] }"
      end

    when /^options=\d+<([^>]+)>/
      options = $~[1].to_s.split(',').collect{|i| i.downcase }

      options.each_index do |oi|
        puts "network.interfaces.#{ i }.options.#{ oi }:str:#{ options[oi] }"
      end

    when /^ether ((?:[0-9a-fA-F]+:){5}[0-9a-fA-F]+)/
      puts "network.interfaces.#{ i }.mac:int:#{ $~[1].to_s.downcase }"

    when /^inet ([^\s]+) netmask ([^\s]+) broadcast (.*)/
      ip = $~[1].to_s
      netmask = $~[2].to_s
      broadcast = $~[3].to_s
      cidr = (netmask.downcase.count('f') * 4)
      netmask = cidr_to_mask(cidr)

      ai = addresses.length

      puts "network.interfaces.#{ i }.addresses.#{ ai }.ip:str:#{ ip }"
      puts "network.interfaces.#{ i }.addresses.#{ ai }.netmask:str:#{ netmask }"
      puts "network.interfaces.#{ i }.addresses.#{ ai }.cidr:int:#{ cidr }"
      puts "network.interfaces.#{ i }.addresses.#{ ai }.broadcast:str:#{ broadcast }"
      puts "network.interfaces.#{ i }.addresses.#{ ai }.family:str:inet4"

      addresses << ip
      global_addresses << ip

    when /^inet6 ([^\s]+) prefixlen (\d+) scopeid 0x(\d+)/
      ip = $~[1].to_s
      cidr = $~[2].to_s
      scopeid = $~[3].to_s

      ai = addresses.length

      puts "network.interfaces.#{ i }.addresses.#{ ai }.ip:str:#{ ip }"
      puts "network.interfaces.#{ i }.addresses.#{ ai }.cidr:int:#{ cidr }"
      puts "network.interfaces.#{ i }.addresses.#{ ai }.scopeid:str:#{ scopeid }"
      puts "network.interfaces.#{ i }.addresses.#{ ai }.family:str:inet6"

      addresses << ip
      global_addresses << ip

    when /^status: (.*)/
      puts "network.interfaces.#{ i }.status:str:#{ $~[1] }"
    end

    # get interface statistics
    phy = 0
    log = 0

    IO.popen("netstat -i -W -b -n -I '#{ interface }'").read.split("\n").each do |line|
      line = line.strip.chomp
      next if line =~ /^Name\s+/

      # aggregate details for the physical interface
      case line
      when /<([^>]+)>/
        name = $~[1].to_s

        ns_interface, mtu, network, mac, ipkt, ierr, idrop, ibytes, opkt, oerr, obytes, coll = line.split(/\s+/)

        puts "network.interfaces.#{ i }.statistics.physical.#{ phy }.id:str:#{ name }"
        puts "network.interfaces.#{ i }.statistics.physical.#{ phy }.mtu:int:#{ mtu.to_i }"
        puts "network.interfaces.#{ i }.statistics.physical.#{ phy }.mac:str:#{ mac }"
        puts "network.interfaces.#{ i }.statistics.physical.#{ phy }.rx.packets:int:#{ ipkt.to_i }"
        puts "network.interfaces.#{ i }.statistics.physical.#{ phy }.rx.errors:int:#{ ierr.to_i }"
        puts "network.interfaces.#{ i }.statistics.physical.#{ phy }.rx.dropped:int:#{ idrop.to_i }"
        puts "network.interfaces.#{ i }.statistics.physical.#{ phy }.rx.bytes:int:#{ ibytes.to_i }"
        puts "network.interfaces.#{ i }.statistics.physical.#{ phy }.tx.packets:int:#{ opkt.to_i }"
        puts "network.interfaces.#{ i }.statistics.physical.#{ phy }.tx.errors:int:#{ oerr.to_i }"
        puts "network.interfaces.#{ i }.statistics.physical.#{ phy }.tx.bytes:int:#{ obytes.to_i }"
        puts "network.interfaces.#{ i }.statistics.physical.#{ phy }.tx.collisions:int:#{ coll.to_i }"

        phy += 1
      else
        ns_interface, _, network, ip, ipkt, _, _, ibytes, opkt, _, obytes, _ = line.split(/\s+/)
        network, cidr = network.split('/', 2)

        puts "network.interfaces.#{ i }.statistics.logical.#{ log }.network:str:#{ network }"
        puts "network.interfaces.#{ i }.statistics.logical.#{ log }.cidr:int:#{ cidr.to_i }"
        puts "network.interfaces.#{ i }.statistics.logical.#{ log }.address:str:#{ ip }"
        puts "network.interfaces.#{ i }.statistics.logical.#{ log }.rx.packets:int:#{ ipkt.to_i }"
        puts "network.interfaces.#{ i }.statistics.logical.#{ log }.rx.bytes:int:#{ ibytes.to_i }"
        puts "network.interfaces.#{ i }.statistics.logical.#{ log }.tx.packets:int:#{ opkt.to_i }"
        puts "network.interfaces.#{ i }.statistics.logical.#{ log }.tx.bytes:int:#{ obytes.to_i }"

        log += 1
      end
    end
  end
end

global_addresses.each_index do |i|
  puts "network.ip.#{ i }:str:#{ global_addresses[i] }"
end

File: shell.d/platform-freebsd/network-interfaces.py
#!/usr/bin/env python2.7
import os
import sys
import re
from subprocess import check_output, CalledProcessError

RX_IFCONFIG_OPTIONS = re.compile('^options=\d+<([^>]+)>')
RX_IFCONFIG_ETHER = re.compile('^ether ((?:[0-9a-fA-F]+:){5}[0-9a-fA-F]+)')
RX_IFCONFIG_INET4 = re.compile('^inet ([^\s]+) netmask ([^\s]+) broadcast (.*)')
RX_IFCONFIG_INET6 = re.compile('^inet6 ([^\s]+) prefixlen (\d+) scopeid 0x(\d+)')
RX_IFCONFIG_STATUS = re.compile('^status: (.*)')
RX_NETSTAT_PHY = re.compile('<([^>]+)>')


def cidr_to_mask(cidr):
    bits = (
        ("1"*int(cidr)) + ("0"*(32-int(cidr)))
    )

    return '.'.join([
        str(int(bits[0:7], 2)),
        str(int(bits[8:15], 2)),
        str(int(bits[16:23], 2)),
        str(int(bits[24:31], 2)),
    ])

try:
    interface_names = []

    # get all interface names
    interface_names = list(set(re.split('\s+', check_output([
        'ifconfig', '-l',
    ]).strip())))

    global_addresses = []

    # for each interface...
    for i, interface in enumerate(interface_names):
        addresses = []
        info = {}

        # get interface details
        for line in check_output(['ifconfig', interface]).split("\n"):
            line = line.strip()

            # -------------------------------------------------------------------------------------
            m = re.match(
                '^{0}:\s+flags=\d+<([^>]+)> metric (\d+) mtu (\d+)$'.format(interface),
                line
            )
            if m is not None:
                flags = [v.lower() for v in str(m.group(1)).split(',')]
                metric = int(m.group(2))
                mtu = int(m.group(3))

                print("network.interfaces.{0}.name:str:{1}".format(i, interface))
                print("network.interfaces.{0}.metric:int:{1}".format(i, metric))
                print("network.interfaces.{0}.mtu:int:{1}".format(i, mtu))

                for fi, flag in enumerate(flags):
                    print("network.interfaces.{0}.flags.{1}:str:{2}".format(i, fi, flag))

                continue

            # -------------------------------------------------------------------------------------
            m = RX_IFCONFIG_OPTIONS.match(line)
            if m is not None:
                options = [v.lower() for v in str(m.group(1)).split(',')]

                for oi, option in enumerate(options):
                    print("network.interfaces.{0}.options.{1}:str:{2}".format(i, oi, option))

                continue

            # -------------------------------------------------------------------------------------
            m = RX_IFCONFIG_ETHER.match(line)
            if m is not None:
                print("network.interfaces.{0}.mac:int:{1}".format(
                    i,
                    str(m.group(1)).lower()
                ))
                continue

            # -------------------------------------------------------------------------------------
            m = RX_IFCONFIG_INET4.match(line)
            if m is not None:
                ip = str(m.group(1))
                netmask = str(m.group(2))
                broadcast = str(m.group(3))
                cidr = (netmask.lower().count('f') * 4)
                netmask = cidr_to_mask(cidr)

                ai = len(addresses)

                print("network.interfaces.{0}.addresses.{1}.ip:str:{2}".format(i, ai, ip))
                print("network.interfaces.{0}.addresses.{1}.netmask:str:{2}".format(
                    i, ai, netmask
                ))
                print("network.interfaces.{0}.addresses.{1}.cidr:int:{2}".format(i, ai, cidr))
                print("network.interfaces.{0}.addresses.{1}.broadcast:str:{2}".format(
                    i, ai, broadcast
                ))
                print("network.interfaces.{0}.addresses.{1}.family:str:inet4".format(i, ai))

                addresses.append(ip)
                global_addresses.append(ip)
                continue

            # -------------------------------------------------------------------------------------
            m = RX_IFCONFIG_INET6.match(line)
            if m is not None:
                ip = str(m.group(1))
                cidr = str(m.group(2))
                scopeid = str(m.group(3))

                ai = len(addresses)

                print("network.interfaces.{0}.addresses.{1}.ip:str:{2}".format(i, ai, ip))
                print("network.interfaces.{0}.addresses.{1}.cidr:int:{2}".format(i, ai, cidr))
                print("network.interfaces.{0}.addresses.{1}.scopeid:str:{2}".format(
                    i, ai, scopeid
                ))
                print("network.interfaces.{0}.addresses.{1}.family:str:inet6".format(i, ai))

                addresses.append(ip)
                global_addresses.append(ip)
                continue

            # -------------------------------------------------------------------------------------
            m = RX_IFCONFIG_INET6.match(line)
            if m is not None:
                print("network.interfaces.{0}.status:str:{1}".format(i, m.group(1)))
                continue

            # get interface statistics
            phy = 0
            log = 0

            for line in check_output([
                'netstat', '-i', '-W', '-b', '-n', '-I', interface,
            ]).split("\n"):
                line = line.strip()

                if line.startswith('Name'):
                    continue

                # aggregate details for the physical interface
                m = RX_NETSTAT_PHY.match(line)
                if m is not None:
                    name = str(m.group(1))

                    parts = line.split()

                    if len(parts) < 12:
                        continue

                    ns_interface = parts[0]
                    mtu = parts[1]
                    network = parts[2]
                    mac = parts[3]
                    ipkt = parts[4]
                    ierr = parts[5]
                    idrop = parts[6]
                    ibytes = parts[7]
                    opkt = parts[8]
                    oerr = parts[9]
                    obytes = parts[10]
                    coll = parts[11]

                    print(
                        "network.interfaces.{0}.statistics.physical.{1}.id:str:{2}".format(
                            i, phy, name
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.physical.{1}.mtu:int:{2}".format(
                            i, phy, int(mtu)
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.physical.{1}.mac:str:{2}".format(
                            i, phy, mac
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.physical.{1}.rx.packets:int:{2}".format(
                            i, phy, int(ipkt)
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.physical.{1}.rx.errors:int:{2}".format(
                            i, phy, int(ierr)
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.physical.{1}.rx.dropped:int:{2}".format(
                            i, phy, int(idrop)
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.physical.{1}.rx.bytes:int:{2}".format(
                            i, phy, int(ibytes)
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.physical.{1}.tx.packets:int:{2}".format(
                            i, phy, int(opkt)
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.physical.{1}.tx.errors:int:{2}".format(
                            i, phy, int(oerr)
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.physical.{1}.tx.bytes:int:{2}".format(
                            i, phy, int(obytes)
                        )
                    )

                    print(
                        (
                            "network.interfaces.{0}.statistics.physical."
                            "{1}.tx.collisions:int:{2}"
                        ).format(
                            i, phy, int(coll)
                        )
                    )

                    phy += 1
                else:
                    parts = line.split()

                    if len(parts) < 12:
                        continue

                    ns_interface = parts[0]
                    network = parts[2]
                    ip = parts[3]
                    ipkt = parts[4]
                    ibytes = parts[7]
                    opkt = parts[8]
                    obytes = parts[10]

                    pair = network.split('/', 1)

                    if len(pair) > 1:
                        network = pair[0]
                        cidr = pair[1]
                    else:
                        network = pair[0]
                        cidr = 32

                    print(
                        "network.interfaces.{0}.statistics.logical.{1}.network:str:{2}".format(
                            i, log, network
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.logical.{1}.cidr:int:{2}".format(
                            i, log, int(cidr)
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.logical.{1}.address:str:{2}".format(
                            i, log, ip
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.logical.{1}.rx.packets:int:{2}".format(
                            i, log, int(ipkt)
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.logical.{1}.rx.bytes:int:{2}".format(
                            i, log, int(ibytes)
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.logical.{1}.tx.packets:int:{2}".format(
                            i, log, int(opkt)
                        )
                    )

                    print(
                        "network.interfaces.{0}.statistics.logical.{1}.tx.bytes:int:{2}".format(
                            i, log, int(obytes)
                        )
                    )

                    log += 1

    for gi, address in enumerate(global_addresses):
        print("network.ip.{0}:str:{1}".format(gi, address))

except CalledProcessError:
    sys.exit(0)
